File: app/src/main/java/com/example/quizzy/domain/Quiz.java
package com.example.quizzy.domain;

import com.google.gson.annotations.SerializedName;

import java.util.List;

import static com.example.quizzy.domain.CachedQuizKt.NOPASSWORD;

public class Quiz {

    @SerializedName("title")
    String title;
    @SerializedName("password")
    String password = NOPASSWORD;
    @SerializedName("duration")
    double duration;
    @SerializedName("startTime")
    long startTime;

    @SerializedName("tags")
    List<String> tags;
    @SerializedName("questions")
    List<Question> questions;
    @SerializedName("responses")
    List<Response> responses;

    public Quiz(String title, String password, double duration, long startTime,
                List<String> tags, List<Question> questions, List<Response> responses) {
        this.title = title;
        this.password = password;
        this.duration = duration;
        this.startTime = startTime;
        this.tags = tags;
        this.questions = questions;
        this.responses = responses;
    }

    public Quiz() {
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public double getDuration() {
        return duration;
    }

    public void setDuration(double duration) {
        this.duration = duration;
    }

    public long getStartTime() {
        return startTime;
    }

    public void setStartTime(long startTime) {
        this.startTime = startTime;
    }

    public List<String> getTags() {
        return tags;
    }

    public void setTags(List<String> tags) {
        this.tags = tags;
    }

    public List<Question> getQuestions() {
        return questions;
    }

    public void setQuestions(List<Question> questions) {
        this.questions = questions;
    }

    public List<Response> getResponses() {
        return responses;
    }

    public void setResponses(List<Response> responses) {
        this.responses = responses;
    }

//    @NonNull
//    @Override
//    public String toString() {
//        String str = title + "\n";
//        for(int i = 0 ; i < tags.size() ; i++){
//            str = str + tags.get(i) + " ";
//        }
//        str = str + "\n";
//        for(int i = 0 ; i < questions.size() ; i++){
//            str = str + questions.get(i).getDescription() + "\n";
//        }
//        return str;
//    }


    @Override
    public String toString() {
        return "Quiz{" +
                "title='" + title + '\'' +
                ", password='" + password + '\'' +
                ", duration=" + duration +
                ", startTime=" + startTime +
                ", tags=" + tags +
                ", questions=" + questions +
                ", responses=" + responses +
                '}';
    }
}

File: app/src/main/java/com/example/quizzy/domain/Quiz.kt
package com.example.quizzy.domain

import androidx.room.*
import com.example.quizzy.keyGen
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.JsonClass
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import java.util.*

const val SINGLE = "SINGLE"
const val MCQ = "MCQ"
const val TEXT = "TEXT"
const val NOPASSWORD = "NO_PASSWORD"
const val PUBLIC = "PUBLIC"
const val PRIVATE = "PRIVATE"

@JsonClass(generateAdapter = true)
@Entity(tableName = "table_quiz")
data class Quiz(
        @PrimaryKey
        val id: String = keyGen(),
        var title: String = "untitled",
        @TypeConverters(QuestionsConverter::class)
        var questions: List<Question> = listOf(),
        @TypeConverters(ResponsesConverter::class)
        var responses: List<Response> = listOf(),
        @TypeConverters(ListConverter::class)
        var tags: List<String> = listOf(),
        var password: String = NOPASSWORD,
        @ColumnInfo(name = "start_time")
        var startTime: Long = 0L,
        var duration: Int = 0
)

@JsonClass(generateAdapter = true)
@Entity(tableName = "table_question")
data class Question(
        @PrimaryKey(autoGenerate = false)
        val serial: Int,
        val description: String,
        val type: String,
        @TypeConverters(ListConverter::class)
        val options: List<String>,
        val marks: Float,
        @TypeConverters(ListConverter::class)
        val answers: List<String> = listOf(),
        @ColumnInfo(name = "image_uri")
        val imageUri: String? = null
)

class QuestionsConverter {

    private val moshi = Moshi.Builder().build()
    private val type = Types.newParameterizedType(List::class.java, Question::class.java)
    private val questionAdapter: JsonAdapter<List<Question>> = moshi.adapter(type)

    @TypeConverter
    fun fromQuestions(questions: List<Question>?): String? {
        return questionAdapter.toJson(questions)
    }

    @TypeConverter
    fun toQuestions(jsonQuestions: String?): List<Question>? {
        return jsonQuestions?.let {  questionAdapter.fromJson(jsonQuestions) }
    }
}

class ListConverter{
    private val moshi = Moshi.Builder().build()
    private val type = Types.newParameterizedType(List::class.java, String::class.java)
    private val listAdapter: JsonAdapter<List<String>> = moshi.adapter(type)

    @TypeConverter
    fun fromList(list: List<String>?) : String? {
        return listAdapter.toJson(list)
    }
    @TypeConverter
    fun toList(jsonList: String?): List<String>? {
        return jsonList?.let { listAdapter.fromJson(jsonList) }
    }
}

@JsonClass(generateAdapter = true)
@Entity(tableName = "table_response")
data class Response(
        val low: Float,
        val high: Float,
        val message: String,
        val imageUri: String? = null,
        @PrimaryKey(autoGenerate = true)
        val id: Long = 0L
)

class ResponsesConverter {
    private val moshi = Moshi.Builder().build()
    private val type = Types.newParameterizedType(List::class.java, Response::class.java)
    private val responsesAdapter: JsonAdapter<List<Response>> = moshi.adapter(type)

    @TypeConverter
    fun fromResponses(responses: List<Response>): String? {
        return responsesAdapter.toJson(responses)
    }
    @TypeConverter
    fun toResponses(jsonResponses: String): List<Response>? {
        return responsesAdapter.fromJson(jsonResponses)
    }
}

class DateConverter {
    @TypeConverter
    fun fromDate(date: Date) : Long? = date.time

    @TypeConverter
    fun toDate(timeInMillis: Long) : Date? = Date(timeInMillis)
}

@Entity(tableName = "table_quiz_item")
data class QuizItem (
        @PrimaryKey
        val id: String,
        val title: String,
        @ColumnInfo(name = "question_count")
        val questionCount: Int,
        val marks: Float,
        @ColumnInfo(name = "start_time")
        val startTime: Long,
        val duration: Int,
        @ColumnInfo(name = "user_count")
        val userCount: Int,
        val tags: List<String>,
        val difficulty: Float,
        val rating: Float,
        val access: String,
        @ColumnInfo(name = "creator_name")
        val creatorName: String,
        @ColumnInfo(name = "creator_image_uri")
        val creatorImageUri: String
)